File: scala/src/main/scala/leetcode/editor/cn/[297]二叉树的序列化与反序列化.scala
package leetcode.editor.cn

object SerializeAndDeserializeBinaryTree {
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for a binary tree node. class TreeNode(var _value: Int) { var value: Int = _value var left: TreeNode = null var right: TreeNode = null }
    */
  import scala.collection.mutable.ListBuffer

  private class Serializer(val root: TreeNode) {
    private val builder = ListBuffer[String]()

    def serialize(): String =
      def impl(node: TreeNode): Unit =
        if node == null then builder.append("#")
        else
          builder.append(node.value.toString)
          impl(node.left)
          impl(node.right)

      impl(root)
      builder.mkString(",")
  }

  private class Deserializer(val encoded: String) {
    private val nodes = ListBuffer[String]()
    nodes.addAll(encoded.split(","))

    def deserialize(): TreeNode =
      if nodes.isEmpty then null
      else
        val nodeVal = nodes.head
        nodes.remove(0)
        if nodeVal == "#" then null
        else
          val node = TreeNode(nodeVal.toInt)
          node.left = deserialize()
          node.right = deserialize()
          node

  }

  class Codec {
    // Encodes a list of strings to a single string.
    def serialize(root: TreeNode): String =
      if root == null then "#"
      else
        Serializer(root).serialize()


    // Decodes a single string to a list of strings.
    def deserialize(data: String): TreeNode = Deserializer(data).deserialize()
  }

  /** Your Codec object will be instantiated and called as such: var ser = new Codec() var deser = new Codec() val s = ser.serialize(root) val ans = deser.deserialize(s)
    */
//leetcode submit region end(Prohibit modification and deletion)

}

File: ts/leetcode/editor/cn/[297]二叉树的序列化与反序列化.ts
import {TreeNode} from "./util.ts";
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

const NULL = "#"

class SerializerBuilder {
    private sb: string[] = []

    constructor(private treeNode: TreeNode) {
    }

    serialize(): string {
        this.serializeBuilder(this.treeNode)
        return this.sb.join(",")
    }

    serializeBuilder(root: TreeNode | null) {
        if (root == null) {
            this.sb.push(`${NULL}`)
        } else {
            this.sb.push(`${root.val}`)
            this.serializeBuilder(root.left)
            this.serializeBuilder(root.right)
        }
    }
}


/*
 * Encodes a tree to a single string.
 */
function serialize(root: TreeNode | null): string {
    if (root == null) return "#"
    const ser = new SerializerBuilder(root)
    return ser.serialize()
}

class Deserializer {
    nodes: string[]

    constructor(encoded: string) {
        this.nodes = encoded.split(",");
    }

    deserialize(): TreeNode | null {
        if (this.nodes.length == 0) return null

        const first = this.nodes[0]
        this.nodes.splice(0,1)

        if (first === NULL) return null

        const root = new TreeNode(parseInt(first))
        root.left = this.deserialize()
        root.right = this.deserialize()

        return root;
    }
}


/*
 * Decodes your encoded data to tree.
 */
function deserialize(data: string): TreeNode | null {
    const des = new Deserializer(data)
    return des.deserialize()
}


/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */
//leetcode submit region end(Prohibit modification and deletion)

const input = new TreeNode(1, new TreeNode(2), new TreeNode(3))
const r = serialize(null)
const n = deserialize(r)
console.log(n)